scheduler/data-flow/src/test/kotlin/io/seldon/dataflow/kafka/PipelineStatusTest.kt

```diff
@@ -0,0 +1,172 @@
+/*
+Copyright (c) 2024 Seldon Technologies Ltd.
+
+Use of this software is governed BY
+(1) the license included in the LICENSE file or
+(2) if the license included in the LICENSE file is the Business Source License 1.1,
+the Change License after the Change Date as each is defined in accordance with the LICENSE file.
+*/
+
+package io.seldon.dataflow.kafka
+
+import org.apache.kafka.streams.KafkaStreams
+import org.junit.jupiter.api.DisplayName
+import org.junit.jupiter.params.ParameterizedTest
+import org.junit.jupiter.params.provider.Arguments
+import org.junit.jupiter.params.provider.Arguments.arguments
+import org.junit.jupiter.params.provider.MethodSource
+import strikt.api.expect
+import strikt.assertions.isEqualTo
+import java.util.stream.Stream
+
+internal class PipelineStatusTest {
+    @DisplayName("Test PipelineStatus States")
+    @ParameterizedTest(name = "{0} expected isActive:{1} | isError:{2}")
+    @MethodSource()
+    fun checkState(
+        testName: String,
+        expectedIsActive: Boolean,
+        expectedIsError: Boolean,
+        state: PipelineStatus,
+    ) {
+        val statusIsActive = state.isActive()
+        val statusIsError = state.isError()
+
+        expect {
+            that(statusIsActive).isEqualTo(expectedIsActive)
+            that(statusIsError).isEqualTo(expectedIsError)
+        }
+    }
+
+    companion object {
+        private const val IS_ACTIVE = true
+        private const val IS_ERROR = true
+
+        @JvmStatic
+        fun checkState(): Stream<Arguments> =
+            Stream.of(
+                arguments(
+                    "StreamStopped(prevState=null)",
+                    !IS_ACTIVE,
+                    !IS_ERROR,
+                    PipelineStatus.StreamStopped(null),
+                ),
+                arguments(
+                    "StreamStopped(prevState=StreamStopping)",
+                    !IS_ACTIVE,
+                    !IS_ERROR,
+                    PipelineStatus.StreamStopped(PipelineStatus.StreamStopping()),
+                ),
+                arguments(
+                    "StreamStopped(prevState=nested StreamStopped without error)",
+                    !IS_ACTIVE,
+                    !IS_ERROR,
+                    PipelineStatus.StreamStopped(
```

Comment:
can we have this condition for real? multiple restarts perhaps to the scheduler?

Reply:
Yes, in the sense that: A pipeline is in state `PipelineStatus.Started()`. It receives a stop, and then it has state `s1=PipelineStatus.StreamStopped(PipelineStatus.Started())`. (I'm simplifying here to show the nesting limitation idea, typically on stop there would be a transition from `Started()` to `StreamStopping()`)

Now, say the pipeline receives another stop. The new state will be initialised as `s2=PipelineStatus.StreamStopped(s1)`. `s2` is essentially what gets tested above. However, because in the implementation of `StreamStopped` we take measures not to deeply nest `StreamStopped` states,
`s2 == s1`